### PR TITLE
Automatically handle Jersey ClientResponse

### PR DESCRIPTION
**Pull request by adreghiciu**
Avoid as much as possible UniformInterfaceException and:
- In case of 404 throw an NotFoundException
- In case of 400 with an ErrorResponse throw a specific exception

Signed-off-by: Alin Dreghiciu adreghiciu@gmail.com

### CODE REVIEW

File: nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        if ( errors != null && !errors.isEmpty() )
+        {
+            final StringBuilder sb = new StringBuilder();
+            for ( final ErrorMessage error : errors )
+            {
+                if ( sb.length() > 0 )
+                {
+                    sb.append( "\n" );
+                }
+                sb.append( error.getMsg() );
```
Comment (nabcos) [inline nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java]: Shouldn't this also mention id when the ID is not "*"?

**Comment by adreghiciu** [comment]
Without this every subsystem impl have to handle this by their own.
More, in most of current subsystem impls we do not handle and that mens that responses remain unconsumed which results in jersey exceptions

File: nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.List;
+
+import org.sonatype.nexus.client.internal.msg.ErrorMessage;
+import org.sonatype.nexus.client.internal.msg.ErrorResponse;
+
+/**
+ * @since 2.3
+ */
+public class BadRequestException
+    extends RuntimeException
```
Comment (cstamas) [inline nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java]: NexusClientException ?

File: nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/NotFoundException.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.client.internal.rest.jersey;
+
+/**
+ * @since 2.3
+ */
+public class NotFoundException
+    extends RuntimeException
```
Comment (cstamas) [inline nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/NotFoundException.java]: NexusClientException ?

**Comment by cstamas** [comment]
Consider using NexusClientException as top level class for exceptions (is also runtime ex).

File: nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java (hunk 1 above)
Comment (adreghiciu) [inline nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java]: +0, how you guys want it. Something like [id] message ?

**Comment by adreghiciu** [comment]
@cstamas c726fa653c850b258aa35d6395834ae1727bcd4e

**Comment by cstamas** [comment]
:kiss:

File: nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java (hunk 1 above)
Comment (nabcos) [inline nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java]: Not sure how generic the message usually is, and I think adding [id] would not hurt.

**Comment by nabcos** [comment]
+1

File: nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java (hunk 1 above)
Comment (adreghiciu) [inline nexus/nexus-client-core/src/main/java/org/sonatype/nexus/client/internal/rest/jersey/BadRequestException.java]: Added id in 4b1af7d39f3ad67409218da5e6a146e256b8d692

**Pull request merged**